### PR TITLE
Do not insert empty blocks on sync Distributed INSERT

### LINKED ISSUE #14571: INSERT to Distributed sent to a shard with zero weight
No rows are written, though. Still, this may be problematic when this insert fails due to the `too many parts` error, which leads to the parent insert also failing. I haven't reproduced this failure exactly (complicated to do on localhost), but seen it happen for a ClickHouse user in Yandex.

Apparently this only happens with `insert_distributed_sync = 1`.

Cluster config:
```
        <weighted>
            <shard>
                <weight>100</weight>
                <replica>
                    <host>127.0.0.2</host>
                    <port>9000</port>
                </replica>
            </shard>
            <shard>
                <weight>0</weight>
                <replica>
                    <host>127.0.0.3</host>
                    <port>9000</port>
                </replica>
            </shard>
            <shard>
                <weight>100</weight>
                <replica>
                    <host>127.0.0.4</host>
                    <port>9000</port>
                </replica>
            </shard>
        </weighted>
```

Queries:

```
create table t_local (a int) engine MergeTree order by (a);

create table t_dist (a int) engine Distributed(weighted, 'default', 't_local', cityHash64(a));

set insert_distributed_sync = 1;

select written_rows, query_id from system.query_log where initial_query_id = '4c93852e-bdcb-436a-9ad3-4953ccedfcdc' and type != 1;
```

In the query log we see that the query is sent to all three shards, including the one with zero weight, but no rows are sent there.
```
SELECT 
    written_rows,
    query_id
FROM system.query_log
WHERE (initial_query_id = '4c93852e-bdcb-436a-9ad3-4953ccedfcdc') AND (type != 1)

┌─written_rows─┬─query_id─────────────────────────────┐
│            0 │ 177294a7-5612-45ef-bb57-2b36b73d709f │
│       499720 │ 02d724ae-5840-4805-9e5f-747a85332545 │
│       500280 │ a4692444-5409-4aec-a821-d9095e912ca1 │
│      1000000 │ 4c93852e-bdcb-436a-9ad3-4953ccedfcdc │
└──────────────┴──────────────────────────────────────┘
```

For more fun, try the `Log` engine for `t_local` -- the insert deadlocks: https://github.com/ClickHouse/ClickHouse/issues/14570

### PR DESCRIPTION
**Pull request by alexey-milovidov**
Changelog category (leave one):
- Performance Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Don't send empty blocks to shards on synchronous INSERT into Distributed table. This closes #14571.

### CODE REVIEW

File: tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.sql [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+SELECT count() FROM local;
+SELECT count() FROM distributed;
+
+TRUNCATE TABLE local;
+INSERT INTO distributed SELECT number FROM numbers(256) WHERE number < 128;
+SELECT count() FROM local;
+SELECT count() FROM distributed;
+
+DROP TABLE local;
+DROP TABLE distributed;
```
Comment (azat) [inline tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.sql:24]: But how does this test ensures that blocks with zero rows will not be sent? It seems that it does not?

Comment (alexey-milovidov) [reply tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.sql:24]: This test does not check it.
It is checked manually with logs.

Comment (alexey-milovidov) [reply tests/queries/0_stateless/01639_distributed_sync_insert_zero_rows.sql:24]: It is only needed to ensure that neither of these cases breaks.

**Comment by alexey-milovidov** [comment]
#18793

**Comment by azat** [comment]
>@alexey-milovidov Arcadia does not support distributed tables

BTW is there any clue (I've looked for `ARCADIA_BUILD`/`grep -v` for `ya.make` - nothing) in the code that will show this?

**Comment by alexey-milovidov** [comment]
No, they just did not setup testing environment correctly (missing some of configs for server to listen other than localhost).

**Pull request merged**